apiserver/observer/recorder.go

```diff
@@ -95,7 +93,93 @@ func (cr *combinedRecorder) HandleReply(req rpc.Request, replyHdr *rpc.Header, b
 	}))
 }
 
-func extractErrors(body interface{}) ([]*auditlog.Error, error) {
-	// TODO(babbageclunk): use reflection to find errors in the response body.
-	return nil, nil
+func extractErrors(body interface{}) []*auditlog.Error {
+	// To find errors in the API responses, we look for a struct where
+	// there is an attribute that is:
+	// * a slice of structs that have an attribute that is *Error or
+	// * a plain old *Error
+	// I thought we'd need to handle a []*Error as well, but it turns
+	// out we don't use that in the API.
+	value := reflect.ValueOf(body)
+	if value.Kind() != reflect.Struct {
+		return nil
+	}
+
+	// Prefer a slice of structs with Errors.
+	for i := 0; i < value.NumField(); i++ {
+		attr := value.Field(i)
+		if errors, ok := tryStructSliceErrors(attr); ok {
+			return convertErrors(errors)
+		}
+	}
+
+	for i := 0; i < value.NumField(); i++ {
+		attr := value.Field(i)
+		if err, ok := tryErrorPointer(attr); ok {
+			return convertErrors([]*params.Error{err})
+		}
+	}
+	return nil
+}
+
+func tryErrorPointer(value reflect.Value) (*params.Error, bool) {
+	if !value.CanInterface() {
+		return nil, false
+	}
+	err, ok := value.Interface().(*params.Error)
+	return err, ok
+}
+
+func tryStructSliceErrors(value reflect.Value) ([]*params.Error, bool) {
+	if value.Kind() != reflect.Slice {
+		return nil, false
+	}
+	itemType := value.Type().Elem()
+	if itemType.Kind() != reflect.Struct {
+		return nil, false
+	}
+	errorField, found := findErrorField(itemType)
+	if !found {
+		return nil, false
+	}
+
+	result := make([]*params.Error, value.Len())
+	for i := 0; i < value.Len(); i++ {
```

Comment:
This will leave nil values in the result slice, is that intended?
I guess so because that will result in a nil auditlog.Error value?

Reply:
Yes, that was intentional - although maybe I can make that clearer if by using append instead.